### PR TITLE
Make sure keywords are single upper words

### PR DESCRIPTION
**Pull request by mvorisek**
Into 1.4.x as this is needed for #106 bugfix.

In this PR, we make sure all "reserved" keywords are single word - this is how SQL specifications (MySQL/MariaDB/SQLite at least) define them.

The "reserved" keywords are special as they need to be always quoted\*. The "reserved" keyword list will be updated in #106 to contain all and only the "reserved" keywords.

\* this is not only helpful for maintaining the lists properly, but also needed for #73 as only "reserved" keywords can be upper cased safely (issue #95).

### CODE REVIEW

**Comment by mvorisek** [comment]
@greg0ire can I ask you for a review?

File: tests/TokenizerTest.php
Comment (greg0ire) [inline tests/TokenizerTest.php]: Instead of painstakingly using reflection to access these 3 properties, you could bind an extra method to the class:

```php
        $assertReservedTopLevelAndNewlineKeywordsAreReserved = Closure::bind(
            function (): void {
                /** @psalm-scope-this Tokenizer */
                $kws = [];

                foreach ([...$this->reservedToplevel, ...$this->reservedNewline] as $v) {
                    foreach (explode(' ', $v) as $kw) {
                        $kws[] = $kw;
                    }
                }

                $kwsDiff = array_diff($kws, $this->reserved);

                TestCase::assertSame([], $kwsDiff);
            },
            new Tokenizer(),
            Tokenizer::class,
        );
        assert($assertReservedTopLevelAndNewlineKeywordsAreReserved instanceof Closure);
        $assertReservedTopLevelAndNewlineKeywordsAreReserved();
```

File: tests/clihighlight.txt
Comment (greg0ire) [inline tests/clihighlight.txt:919]: That line contains `[0m` three times in a row… that's not valid, is it?

Comment (mvorisek) [reply tests/clihighlight.txt:919]: IDK, in any case, this is unrelated to this PR, previously two reserved words were easily tokenized as well.

File: tests/TokenizerTest.php
Comment (mvorisek) [reply tests/TokenizerTest.php]: I know `Closure::bind` but as you mentioned in the code example Psalm needs the `@var` comment as well so I definitely do not find it better.

Once https://github.com/phpstan/phpstan/issues/3913#issuecomment-2117363294 will be implemented the current solution should be best and no `@var` will be needed. Psalm does not need `@var` for the current solution already.

File: tests/clihighlight.txt
Comment (greg0ire) [reply tests/clihighlight.txt:919]: Previously, there wasn't two consecutive `[0m` without a `[37m` in between.

File: tests/TokenizerTest.php
Comment (greg0ire) [reply tests/TokenizerTest.php]: > but as you mentioned in the code example Psalm needs the `@var` comment as well so I definitely do not find it better.

What are you talking about? My code sample does not contain any occurrence of `@var`

Comment (mvorisek) [reply tests/TokenizerTest.php]: Sorry, I meant "phpdoc comment" - `@psalm-scope-this`.

File: tests/clihighlight.txt
Comment (mvorisek) [reply tests/clihighlight.txt:919]: To answer your initial question, all should be fine, `[37m` sets "light gray color" and `[0m` resets it.

The `CURRENT` keyword was not present before but is reserved and it will be fixed in https://github.com/doctrine/sql-formatter/pull/106. Can this PR be merged as is or do you want me to port the `CURRENT` keyword into this PR?

Currently this is printed as:
image
(the colors are close, monitor with good colors is needed)

I prefer to fix this in the #106 PR as other keywords (like `NO`, `OTHERS`, ...) were/are not highlighted as keywords either.

File: tests/TokenizerTest.php
Comment (greg0ire) [reply tests/TokenizerTest.php]: Oh I see… Even with `@var` gone, I think I still prefer my solution:
- no reflection
- fewer variables

File: tests/clihighlight.txt
Comment (greg0ire) [reply tests/clihighlight.txt:919]: Fine I guess.

File: tests/TokenizerTest.php
Comment (mvorisek) [reply tests/TokenizerTest.php]: In #106 the data are even more asserted, so I moved the reflection part into a private getter. Is that fine and can thsi PR be merged?

**Comment by greg0ire** [comment]
> Changes are explained by added tests & commit messages.

Changes are described, but they are not explained by commit messages (maybe by tests, but that's not sufficient I'm afraid). For instance, what makes you think keywords appearing in `reservedTopLevel` must also appear in `reserved`?

**Comment by mvorisek** [comment]
> what makes you think keywords appearing in `reservedTopLevel` must also appear in `reserved`?

"reserved" words are special - they needs to be quoted even if used on places where only identifier is possible (ex `select x as AND` imply a parse error for both MySQL and SQLite). This PR is cleanup of current code before #106 is landed.

In #106, "reserved" words will be asserted by complete lists.

**Comment by greg0ire** [comment]
I'm asking because `reserved` and derived properties seems mostly used inside this code block, where being part of `reserved` does not seem to do much if you are already part of `reservedTopLevel` or `reservedNewline`:

https://github.com/doctrine/sql-formatter/blob/6eb85e0b18135baadd61a7f80b8fa668519b6cd5/src/Tokenizer.php#L887-L932

Is there a part of the code where being part of `reserved` matters even if you are already part of the two other properties?

**Comment by mvorisek** [comment]
> I'm asking because `reserved` and derived properties seems mostly used inside this code block, where being part of `reserved` does not seem to do much if you are already part of `reservedTopLevel` or `reservedNewline`:

I have now removed the test I added testing if all new line/top level keywords are also reserved, as in theory, this does not have to be true.

Currently the lists are in quite bad shape, for ex. `WITH` keyword is already present in `reservedToplevel` but also in `reserved`.

Reserved keywords should be maintained however on its own and they will be asserted in #106, the reason is "reserved" keyworda are special as they must be always quoted - and they will be needed to address #95. "reserved top level" and "reserved new line" are just bad historical names.

**Comment by mvorisek** [comment]
@greg0ire I have now amended the commits with explanation messages.

**Comment by mvorisek** [comment]
@greg0ire I would like to continue with #106, please, can you review this bugfix PR?

**Review by greg0ire — APPROVED** [review]

File: tests/TokenizerTest.php
Comment (SenseException) [inline tests/TokenizerTest.php]: What is this method supposed to test? This is checking the format of the reserved keywords.

Comment (mvorisek) [reply tests/TokenizerTest.php]: All reserved keywords must be a single word in upper case (so we can rely on it when comparing againts other UC string).

File: tests/TokenizerTest.php
Comment (goetas) [inline tests/TokenizerTest.php]: I think that this test is testing something "too" internal. they way how properties are used in the class should not be IMO really tested. changing the implementation at this level of depth should not really impact tests IMO

File: src/Tokenizer.php
Comment (goetas) [inline src/Tokenizer.php]: i'm not really sure about this change.... `ORDER BY` is a keyword. `ORDER` by it self in most of sql dialects means nothing. (same for all the other "composite keywords we had").

If we were to say that keywords are uppercase and just "single words", this library would be much simpler....

**Review by goetas — COMMENTED** [review]
Overall I do not see the need for this change (except of we want to keep the array of keywords alphabetically sorted)

File: src/Tokenizer.php
Comment (mvorisek) [reply src/Tokenizer.php]: Here is MySQL spec - https://dev.mysql.com/doc/refman/8.4/en/keywords.html

`ORDER` and `BY` are "reserved" keywords and must be specified separately.

"reserved" keywords are special as they cannot be used in SQL queries without quotes - ex: https://dbfiddle.uk/_IuDbiFR

This is also why we have the `testKeywordsReservedAreUpperCasedAndWithoutWhitespace` test :)

File: tests/TokenizerTest.php
Comment (mvorisek) [reply tests/TokenizerTest.php]: explained in https://github.com/doctrine/sql-formatter/pull/107#discussion_r1609798033

File: src/Tokenizer.php
Comment (goetas) [reply src/Tokenizer.php]: how would  the query `select 'x' AS order` be formatted by this library before and after your changes?

Comment (mvorisek) [reply src/Tokenizer.php:322]: Perfectly, as it is invalid query as shown in the online repro :)

Comment (derrabus) [reply src/Tokenizer.php]: > "reserved" keywords are special as they cannot be used in SQL queries without quotes

That may be the case. However, it is not the objective of this library to detect syntax errors, but to format a given SQL query for rendering it e.g. in the browser.

Comment (mvorisek) [reply src/Tokenizer.php]: At first, it is invalid query.

But to make you satisfied, it will be formatted exactly the same as before. This PR fixes "reserved" keywords.

"reservedToplevel" "possible multi-word keywords" are the same as before. Please ignore the "reserved" prefix of "reservedToplevel" private property as the "reserved" there is misleading but the naming was introduced years ago wrongly...

"reserved" keywords needs to match the SQL spec to be able to rely they needs to be quoted - and prevent bugs like #95.

Comment (mvorisek) [reply src/Tokenizer.php]: @goetas @derrabus does it make sense to you and can I get your approval so I can continue with #106 where the "reserved" keywords will be extended and fixed to match strictly the spec?

Comment (goetas) [reply src/Tokenizer.php]: Could you please answer to https://github.com/doctrine/sql-formatter/pull/107#discussion_r1609804049?

Comment (mvorisek) [reply src/Tokenizer.php]: @goetas answered in https://github.com/doctrine/sql-formatter/pull/107#discussion_r1609812253 - the answer is - exactly the same as before.

**Comment by goetas** [comment]
@mvorisek just some feedback, The commit https://github.com/doctrine/sql-formatter/pull/107/commits/2f8430868815ecebe6e45cf736972a99a8405040 makes it very hard to review your changes... 

i see a lot of small changes even in the other two commits about sorting variables in the arrays that honestly add little value and makes the reviewer work much harder

**Comment by goetas** [comment]
Would it be possible to isolate the "real" changes you did from the "esthetical" changes?

**Comment by mvorisek** [comment]
@goetas sure - https://github.com/doctrine/sql-formatter/compare/2f843086...54731080?expand=1

**Comment by goetas** [comment]
`CHARACTER SET` seems removed and I do not see `CHARACTER` in the list of keywords.

I'm undecided on this PR but since technically it does not make the project "worse" do not want to block it and will leave it to @derrabus  and @greg0ire  to decide.

I personally do not like the last commit that shuffles a lot of line of code... and sooner or later someone else will add a keyword not following alphabetical order  (we have no CS checks to ensure it)

**Review by goetas — DISMISSED** [review]

**Review by greg0ire — CHANGES_REQUESTED** [review]
Since every reviewer had issues with at least one of the commits, I suggest you do one PR per commit as hinted by @goetas in https://github.com/doctrine/sql-formatter/pull/107#issuecomment-2125002955

**Comment by mvorisek** [comment]
@greg0ire the sorting PR (#117) was merged, can this PR be merged as well?

**Comment by greg0ire** [comment]
I don't know if @goetas @derrabus and @SenseException are satisfied with your answers. Since you can't get approvals, I am closing this PR and I suggest you do one PR per commit, so that there is more hope to close open discussions, and so that we do not have to look at too many other PRs and issue to understand what you are trying to do.

In these new PRs, try to be less pushy, and to explains things better, with clearer sentences. Take your time. I'll give you an example:

>  to be able to use faster CS comparison, they are always CI

Here it is unclear if you are referring to things before or after your changes, and it is unclear what you are referring to with "they".

**Comment by mvorisek** [comment]
@greg0ire please reopen this PR and I will extract the 2nd commit into another PR. Thank you.

**Comment by mvorisek** [comment]
PR is done and I did my best to explain the change in the PR description.

File: src/Tokenizer.php
Comment (greg0ire) [inline src/Tokenizer.php:185]: I think I might not be understanding what you are doing here… shouldn't you be adding back `NO` as well as `OTHERS` to this list?

Comment (mvorisek) [reply src/Tokenizer.php:185]: `NO` or `OTHERS` must not be added into "reserved" keywords as both are not "reserved" by SQL/MySQL spec.

Comment (greg0ire) [reply src/Tokenizer.php:185]: > Make sure keywords are single upper words

Your commit message does not mention that at any point.

Comment (mvorisek) [reply src/Tokenizer.php:185]: The PR description does. If the commit message should be amended, please tell me what text to put there.

Comment (greg0ire) [reply src/Tokenizer.php:185]: Well maybe it should be a separate commit, since it's an entirely separate concern.

Comment (mvorisek) [reply src/Tokenizer.php:185]: We need to end up exactly with the changes as this PR does - please tell me exactly what to do, am I right you want separate commit for `'NO OTHERS'` keyword removal? Anything else?

Comment (greg0ire) [reply src/Tokenizer.php:185]: 2 commits:
- one that removes `NO OTHERS` and explains why, and also takes a guess at why `NO OTHERS` was present in this list in the first place
- another that is about making sure we deal with single words, and also explains why that is necessary.

Comment (mvorisek) [reply src/Tokenizer.php:185]: Done and thank you.

Comment (greg0ire) [reply src/Tokenizer.php:185]: Why do you think `NO OTHERS` was present in the list in the first place? You cite MySQL grammar, but this library is not solely about MySQL, is it?

Comment (mvorisek) [reply src/Tokenizer.php:185]: > Why do you think `NO OTHERS` was present in the list in the first place?

See https://github.com/search?q=repo%3Amysql%2Fmysql-server+%22NO+OTHERS%22&type=code MySQL tests for usecases and https://github.com/antlr/grammars-v4/blob/4264f19a7d/sql/mysql/Oracle/MySQLParser.g4#L1208 for full grammar.

It is however not a "reserved" keyword because of two reasons. Keyword must be a single word. And must be listed as reserved in MySQL or MariaDB docs.

> You cite MySQL grammar, but this library is not solely about MySQL, is it?

The jdorm library was for MySQL dialect only. In #106 many keywords will be added as the current list is far from complete.

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @mvorisek !

**Pull request merged**
